sources/pg_replication/helpers.py

```diff
@@ -0,0 +1,757 @@
+                "Truncate replication messages are ignored."
+            )
+
+    def process_relation(self, decoded_msg: Relation) -> None:
+        """Processes a replication message of type Relation.
+
+        Stores table schema information from Relation message in object state.
+        """
+        # store table schema information
+        columns = {c.name: _to_dlt_column_schema(c) for c in decoded_msg.columns}
+        self.last_table_schema[decoded_msg.relation_id] = {
+            "name": decoded_msg.relation_name,
+            "columns": columns,
+        }
+
+    def process_change(
+        self, decoded_msg: Union[Insert, Update, Delete], msg_start_lsn: int
+    ) -> None:
+        """Processes replication message of type Insert, Update, or Delete.
+
+        Adds data item for inserted/updated/deleted record to instance attribute.
+        """
+        if isinstance(decoded_msg, (Insert, Update)):
+            column_data = decoded_msg.new_tuple.column_data
+        elif isinstance(decoded_msg, Delete):
+            column_data = decoded_msg.old_tuple.column_data
+        table_name = self.last_table_schema[decoded_msg.relation_id]["name"]
+        data_item = self.gen_data_item(
+            table_name=self.last_table_schema[decoded_msg.relation_id]["name"],
+            data=column_data,
+            column_schema=self.last_table_schema[decoded_msg.relation_id]["columns"],
+            lsn=msg_start_lsn,
+            commit_ts=self.last_commit_ts,
+            for_delete=isinstance(decoded_msg, Delete),
+            include_columns=None
+            if self.include_columns is None
+            else self.include_columns.get(table_name),
+            column_hints=None if self.columns is None else self.columns.get(table_name),
+        )
+        self.data_items.append(data_item)
+
+    @staticmethod
+    def gen_data_item(
+        table_name: str,
+        data: List[ColumnData],
+        column_schema: TTableSchemaColumns,
+        lsn: int,
+        commit_ts: pendulum.DateTime,
+        for_delete: bool,
+        include_columns: Optional[Sequence[str]] = None,
+        column_hints: TTableHintTemplate[TAnySchemaColumns] = None,
+    ) -> TDataItem:
+        """Generates data item from replication message data and corresponding metadata."""
+        pairs = zip(column_schema.values(), data)
+        if include_columns is not None:
+            pairs = [(schema, data) for (schema, data) in pairs if schema["name"] in include_columns]  # type: ignore[assignment]
+        data_item = {
+            schema["name"]: _to_dlt_val(
+                val=data.col_data,
+                data_type=schema["data_type"],
+                byte1=data.col_data_category,
+                for_delete=for_delete,
+            )
+            for (schema, data) in pairs
+        }
+        data_item["lsn"] = lsn
+        if for_delete:
+            data_item["deleted_ts"] = commit_ts
+        return dlt.mark.with_hints(
```

Comment:
this is all correct but will be very slow. what happens currently is
- you yield items one by one
- each item has full hints with it.
this will generate a lot of unnecessary processing in the pipeline.

what you should do instead (I think it is not very complicated):
- buffer data items per table name to be able to yield them together
- keep schema changes per table and yield them once
you yield stuff when we reach batch size:
- for each table and schema change
`yield dlt.mark.with_hints([], dlt.mark.make_hints(...)` - you can yield empty list just to update hints
- for each table and data items
`yield dlt.mark.with_table_name(data_items[table_name], table_name)`
where `data_items[table_name]` is a list of data items collected for a particular table

the speedup will be significant. and this code structure also makes easier to yield arrow tables if we want to be even faster
tell me WDYT

Reply:
Much better indeed! I started building this, but faced the "hint accumulation [issue](https://github.com/dlt-hub/dlt/issues/1105)" that's blocking progress. This is on hold until that issue has been completed.

Reply:
This has now been implemented in https://github.com/dlt-hub/verified-sources/pull/392/commits/5636e078c9f4b0085549d13cdc29bb5e17224eae.